faunadb-scala/src/test/scala/faunadb/ClientSpec.scala

```diff
@@ -1616,9 +1627,362 @@ class ClientSpec
+        t3("event", "prev").toOpt.isEmpty shouldBe true
+        t3("event", "diff").toOpt.isEmpty shouldBe true
+      case _ =>
+        fail("expected 4 events")
+    }
+  }
+
+  it should "stream on document reference with opt-in fields" in {
+    val createdDoc = client.query(Create(Collection("spells"), Obj("data" -> Obj("testField" -> "testValue0")))).futureValue
+    val docRef = createdDoc(RefField)
+    val publisherValue = client.stream(docRef, List(DocumentField, PrevField, DiffField)).futureValue
+    val events = testSubscriber(4, publisherValue)
+
+    // push 3 updates
+    client.query(Update(docRef, Obj("data" -> Obj("testField" -> "testValue1")))).futureValue
+    client.query(Update(docRef, Obj("data" -> Obj("testField" -> "testValue2")))).futureValue
+    client.query(Update(docRef, Obj("data" -> Obj("testField" -> "testValue3")))).futureValue
+
+    // assertion 4 events (start + 3 updates)
+    events.futureValue match {
+      case startEvent :: t1 :: t2 :: t3 :: Nil =>
+        startEvent("type").get shouldBe StringV("start")
+        startEvent("event").toOpt.isDefined shouldBe true
+
+        t1("type").get shouldBe StringV("version")
+        t1("event", "document", "data").get shouldBe ObjectV("testField" -> StringV("testValue1"))
+        t1("event", "prev", "data").get shouldBe ObjectV("testField" -> StringV("testValue0"))
+        t1("event", "diff", "data").get shouldBe ObjectV("testField" -> StringV("testValue1"))
+
+        t2("type").get shouldBe StringV("version")
+        t2("event", "document", "data").get shouldBe ObjectV("testField" -> StringV("testValue2"))
+        t2("event", "prev", "data").get shouldBe ObjectV("testField" -> StringV("testValue1"))
+        t2("event", "diff", "data").get shouldBe ObjectV("testField" -> StringV("testValue2"))
+
+        t3("type").get shouldBe StringV("version")
+        t3("event", "document", "data").get shouldBe ObjectV("testField" -> StringV("testValue3"))
+        t3("event", "prev", "data").get shouldBe ObjectV("testField" -> StringV("testValue2"))
+        t3("event", "diff", "data").get shouldBe ObjectV("testField" -> StringV("testValue3"))
+      case _ =>
+        fail("expected 4 events")
+    }
+  }
+
+  it should "stream updates last seen transaction time" in {
+    val createdDoc = client.query(Create(Collection("spells"), Obj("credentials" -> Obj("password" -> "abcdefg")))).futureValue
+    val docRef = createdDoc(RefField)
+    val publisherValue = client.stream(docRef).futureValue
+
+    val capturedEventsP = Promise[List[Value]]
+
+    val valueSubscriber = new Flow.Subscriber[Value] {
+      var subscription: Flow.Subscription = null
+      val captured = new ConcurrentLinkedQueue[Value]
+
+      override def onSubscribe(s: Flow.Subscription): Unit = {
+        subscription = s
+        subscription.request(1)
+      }
+
+      override def onNext(v: Value): Unit = {
+        if (v("txn").to[Long].get <= client.lastTxnTime) {
```

Comment:
I think you can test for `<` only. If you block between requests, the next txn must have a greater txn timestamp.

Reply:
I am not sure about this one :thinking: 

Without concurrent handling of events, the client's txn is equal to the last event's txn?

Also changing `<=` to `<` breaks my test.

Reply:
Do you update the client's txn time before forwarding events downstream? If so, that'd explain the need for equality check, which would then be okay.

Reply:
Yes, I do update the transaction time before pushing the events downstream :+1: